### ISSUE TITLE
Consider removing using non-unicode Encoding in the library, leaving the use of FallbackEncoding 

### ISSUE THREAD

**Issue by rstm-sf** [ExcelDataReader/ExcelDataReader#456]
Hello!

Consider removing using non-unicode ```Encoding``` in the library, leaving the use of ```FallbackEncoding```.

Or add use ```CodePagesEncodingProvider.Instance.GetEncoding(...)```~~, or register ```CodePagesEncodingProvider.Instance```~~ (adding target netcoreapp3_0?)

The background can be read here #445

**Comment by appel1** [comment]
I don't think we should implicitly change global state by registering the encoding provider.  Nor should we use the provider directly since it'd be confusing in projects that target multiple versions, or if using a different provider depending on target platform. 

Perhaps we could lazy load the default value for FallbackEncoding so we don't always require 1252 to be available. We should check if that improves things before making any change though. If we in most cases need 1252 anyways it would be a useless change.

**Comment by rstm-sf** [comment]
> I don't think we should implicitly change global state by registering the encoding provider.

If you register on side the library, it will only be on the library side. The only difference is that you will need to use an additional codepage library and additional encoding provider in the library :)

**Comment by rstm-sf** [comment]
I only suggest simplifying the use of the library to users of the .NET Core platform (in fact, this is the main central audience)

**Comment by appel1** [comment]
I don't understand, calling `System.Text.Encoding.RegisterProvider` registers it globally for the whole process.

**Comment by rstm-sf** [comment]
I checked it several times on netcore 3.0. But I’m ready to check again, having already created a demo solution, to be sure that I was not mistaken :)
(Sorry, but I can do it in at least 2 days)

**Comment by andersnm** [comment]
I think the options are:

1. The library can call `RegisterProvider` itself on netstandard/netcoreapp targets (simple to do, likely OK 99.9% of the time)
2. The library can use `CodePagesEncodingProvider.Instance` directly on netstandard/netcoreapp targets and retain current approach on net20/net45 targets (minor refactoring task, but not sure if this works with XmlReader encodings)

For now I'd rather keep this issue in the backlog as things are currently relatively simple and well understood. I worry any changes could have side effects down the line across runtimes, versions etc which would need further investigation and documentation (= time). 

To add a bit about the state of encodings:

In most cases ExcelDataReader uses UTF8 (default XML encoding in XLSX) or UTF16 (only supported encoding in BIFF8 XLS). But BIFF2-5 XLSes always require some non-unicode encoding, out of which 1252 is presumably the most common one: older Excels running on English DOS/Windows use this. And third party XLSX libraries could also write XML using non-unicode encodings.

Thus removing non-Unicode encodings is out of the question, since these are an integral part of the Excel file format.

The reason for initializing FallbackEncoding eagerly was specifically to crash early, forcing developers to register the codepage providers and avoiding a situation where things may seem to work initially during testing/development only to blow up in production when users try to load certain files.

(also, `RegisterProvider` uses a static array to hold the providers, which should be global to the process. I don't think .NET has any concept of "assembly-local storage". It is however safe to call several times with the same provider)

**Comment by appel1** [comment]
I don't think any of those options are good and if those are our only options we should opt to leave things as they are.

**Comment by rstm-sf** [comment]
> I checked it several times on netcore 3.0. But I’m ready to check again, having already created a demo solution, to be sure that I was not mistaken :)
> (Sorry, but I can do it in at least 2 days)

Done: rstm-sf/DemoLocalRegisterEncodingProvider#2

**Comment by andersnm** [comment]
@rstm-sf Thanks, but alas the tests depend on run order. The tests run alphabetically and fail if you rename them to f.ex TestA and TestB

**Comment by rstm-sf** [comment]
> @rstm-sf Thanks, but alas the tests depend on run order. The tests run alphabetically and fail if you rename them to f.ex TestA and TestB

Wow! Thank you for pointing out the error. I did not consider it

**Comment by rstm-sf** [comment]
So it remains to be used as ```CodePagesEncodingProvider.Instance.GetEncoding(...)```

**Comment by rstm-sf** [comment]
> I don't think any of those options are good and if those are our only options we should opt to leave things as they are.

You can try to use a helper like this

```c#
try
{
    return Encoding.GetEncoding(encodingName);
}
catch (ArgumentException) // unsupported name
{
#if NETSTANDARD && !NETSTANDARD1_0
    return CodePagesEncodingProvider.Instance.GetEncoding(encodingName);
#else
    return null;
#endif
}
```

**Comment by awright18** [comment]
This is unusable on any .NET Core version including .NET 5 without some kind of change.  I tried to construct my own configuration and pass the encoding provider but the default Encoding.GetEncoding(encodingName); gets called before anything can set FallbackEncoding. It would be nice if the ExcelReaderConfiguration had a constructor that allowed passing in the FallbackEncoding to avoid this problem. Otherwise I'm not sure this will work with any .NET Core/5+ Applications.
